storage/namespace.go

```diff
@@ -143,6 +143,69 @@ func (n *dbNamespace) NumSeries() int64 {
 	return count
 }
 
+func (n *dbNamespace) Shards() []Shard {
+	n.RLock()
+	shards := n.shardSet.Shards()
+	databaseShards := make([]Shard, len(shards))
+	for i, shard := range shards {
+		databaseShards[i] = n.shards[shard]
+	}
+	n.RUnlock()
+	return databaseShards
+}
+
+func (n *dbNamespace) AssignShardSet(shardSet sharding.ShardSet) {
+	var (
+		incoming = make(map[uint32]struct{})
+		existing []databaseShard
+		closing  []databaseShard
+	)
+	for _, shard := range shardSet.Shards() {
+		incoming[shard] = struct{}{}
+	}
+
+	n.Lock()
+	existing = n.shards
+	for _, shard := range existing {
+		if shard == nil {
+			continue
+		}
+		if _, ok := incoming[shard.ID()]; !ok {
+			closing = append(closing, shard)
+		}
+	}
+	n.shardSet = shardSet
+	n.shards = make([]databaseShard, n.shardSet.Max()+1)
+	for _, shard := range n.shardSet.Shards() {
+		if int(shard) < len(existing) && existing[shard] != nil {
+			n.shards[shard] = existing[shard]
+		} else {
+			n.shards[shard] = newDatabaseShard(n.id, shard, n.increasingIndex,
+				n.writeCommitLogFn, n.nopts.NeedsBootstrap(), n.sopts)
+		}
+	}
+	n.Unlock()
+
+	// NB(r): There is a shard close deadline that controls how fast each
```

Comment:
Soft deadline.  If we don't meet it we'll just take a little longer than we asked it to.